### PR TITLE
KNOX-2809 - Fix dispatch ha-class for HDFSUI high available.

### PR DESCRIPTION
**Pull request by Flamingo93**
## What changes were proposed in this pull request?
@risdenk use `org.apache.knox.gateway.dispatch.URLDecodingDispatch` for HDFSUI dispatch ha-class  in https://github.com/apache/knox/pull/103. But `URLDecodingDispatch` didn't do failover request when request failed. 

I recommend using `org.apache.knox.gateway.hdfs.dispatch.HdfsUIHaDispatch` for this, but it  it lack URL decoding what  `URLDecodingDispatch` do. So I override `getDispatchUrl` method in `HdfsUIHaDispatch`, decode url just like `URLDecodingDispatch`.

## How was this patch tested?

- mvn clean verify -Ppackage,release
- Checked that HDFSUI service definition works with HaProvider on cluster

image

### CODE REVIEW

**Comment by Flamingo93** [comment]
@smolnar82 Travis CI is failed, and I can't load the detail of failed job. Can you help me to fix this?

**Comment by smolnar82** [comment]
@Flamingo93 - Travis CI failed for the ARM64 architecture and it's a known issue by the Knox team. For now, we can disregard that failure.

**Comment by Flamingo93** [comment]
Hi @risdenk @lmccay @moresandeep @pzampino @smolnar82 Could you please take a look? :)

**Review by smolnar82 — CHANGES_REQUESTED** [review]

File: gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/HdfsUIHaDispatch.java [hunk 1]
```diff
@@ -26,6 +30,26 @@ public HdfsUIHaDispatch() throws ServletException {
...
+
+    final StringBuilder str = new StringBuilder(decoded);
+    final String query = request.getQueryString();
+    if ( query != null ) {
+      str.append('?');
+      str.append(query);
+    }
+    return URI.create(str.toString());
+  }
+
```
[the remark this answers is not in the document]
Comment (smolnar82) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/HdfsUIHaDispatch.java:52]: @Flamingo93 - any progress on this thread? Please address my above review comment so that we can continue merging this PR. Thanks!

[the remark this answers is not in the document]
Comment (Flamingo93) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/HdfsUIHaDispatch.java]: @smolnar82 sorry for late. 
I create a util class for dispatch url processing as your recommendation. Thanks for your recommendation and patience.

File: gateway-spi/src/main/java/org/apache/knox/gateway/util/URLUtils.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+public class URLUtils {
+
+  /**
+   * A method that decode url and concat query params, return result as URI.
+   *
+   * @param encoded encoded url
+   * @param queryStr query params
+   * @return decoded URI
+   */
+  public static URI getDecodeUri(String encoded, String queryStr) {
```
Comment (smolnar82) [inline gateway-spi/src/main/java/org/apache/knox/gateway/util/URLUtils.java:32]: nit: I think there is a typo here; it should be `getDecodedUri` (note the `d` at the end of `Decoded`)or the method name should be changed to a verb: `decodeUri`

File: gateway-spi/src/main/java/org/apache/knox/gateway/util/URLUtils.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    String decoded;
+    try {
+      decoded = URLDecoder.decode(encoded, StandardCharsets.UTF_8.name());
+    } catch (final Exception e) {
+      /* fall back in case of exception */
+      decoded = encoded;
+    }
+
+    final StringBuilder str = new StringBuilder(decoded);
+    if (queryStr != null) {
```
Comment (smolnar82) [inline gateway-spi/src/main/java/org/apache/knox/gateway/util/URLUtils.java:42]: You may want to check for being `not blank` here (`StringUtils.isNotBlank(queryStr)`). Even if this is not null, it may be a blank string that makes no sense to append at the end of the decoded URI.

File: gateway-spi/src/test/java/org/apache/knox/gateway/util/URLUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ */
+package org.apache.knox.gateway.util;
+
+import org.junit.Test;
+
+import java.net.URI;
+
+import static org.junit.Assert.assertEquals;
+
+public class URLUtilsTest {
```
Comment (smolnar82) [inline gateway-spi/src/test/java/org/apache/knox/gateway/util/URLUtilsTest.java:25]: Please add other test cases with null and empty query strings.

**Review by smolnar82 — CHANGES_REQUESTED** [review]
Just minor changes; please address my comments and let's merge this :)

File: gateway-spi/src/test/java/org/apache/knox/gateway/util/URLUtilsTest.java (hunk 4 above)
Comment (pzampino) [reply gateway-spi/src/test/java/org/apache/knox/gateway/util/URLUtilsTest.java:25]: +1
